core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1349,6 +1339,48 @@ class DAGScheduler(
               s"longer running")
           }
 
+          if (mapStage.rdd.isBarrier()) {
+            // Mark all the map as broken in the map stage, to ensure retry all the tasks on
+            // resubmitted stage attempt.
+            mapOutputTracker.unregisterAllMapOutput(shuffleId)
+          } else if (mapId != -1) {
+            // Mark the map whose fetch failed as broken in the map stage
+            mapOutputTracker.unregisterMapOutput(shuffleId, mapId, bmAddress)
+          }
+
+          if (failedStage.rdd.isBarrier()) {
+            failedStage match {
+              case mapStage: ShuffleMapStage =>
+                // Mark all the map as broken in the map stage, to ensure retry all the tasks on
+                // resubmitted stage attempt.
+                mapOutputTracker.unregisterAllMapOutput(mapStage.shuffleDep.shuffleId)
+
+              case resultStage: ResultStage =>
+                // Mark all the partitions of the result stage to be not finished, to ensure retry
+                // all the tasks on resubmitted stage attempt.
+                resultStage.activeJob.map(_.markAllPartitionsAsUnfinished())
+            }
+          }
+
+          // TODO: mark the executor as failed only if there were lots of fetch failures on it
+          if (bmAddress != null) {
```

Comment:
why move this before the `if (shouldAbortStage) { ...`?

Reply:
IIUC previously there was a race condition, that you first post the `ResubmitFailedStages` message then maybe unregister outputs, if the `ResubmitFailedStages` is handled quick enough then you may first try to submit missing tasks then unregister some shuffle blocks. This code change fixes the issue. Maybe this worth being handled separately?

Reply:
I think so